File: api/threats.py
"""Le relevé public des menaces, par **ligne** et pour un profil donné (`L3`).

Distinct de `/v1/triage`, et il faut dire pourquoi : le diagnostic demande une adresse
et capture un lead, ce qui est un arbitrage commercial assumé. Le relevé, lui, alimente
une section d'une page publique où l'on coche un profil pour voir ce qui change. Le
faire passer par `/v1/triage` obligerait à donner son adresse pour lire une liste de
menaces, ce qui n'est ni nécessaire ni défendable.

D'où : `GET`, aucune écriture, aucune donnée personnelle, aucune base.

**Pourquoi une route et pas un simple artefact statique.** Le regroupement en lignes
est publié tel quel dans `frontend/lib/generated/threat-rows.json` : il ne dépend
d'aucun visiteur. Mais l'applicabilité, le plafond de profil (`FR-174`) et la famille
retenue en dépendent, et ce sont eux qui portent les chiffres. Les recalculer en
TypeScript donnerait un second moteur ; le premier désaccord entre les deux se lirait
sur une page commerciale. Le moteur reste `core.triage`, ici comme ailleurs.

Fail-closed : sans carte publiée, on répond 503 plutôt qu'une liste vide. Une liste
vide se lit « aucune menace », qui est la plus mauvaise des réponses fausses.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request, status

from api.ratelimit import limiter, threats_rate_limit
from core import triage

router = APIRouter(prefix="/v1/threats", tags=["threats"])

_MAP = Path(__file__).resolve().parent.parent / "coverage" / "map.json"

#: Borne de longueur sur le paramètre, avant même de le découper (`CLAUDE.md` §4.9).
#: Six profils de trois à quatre caractères plus leurs virgules tiennent en 30.
_MAX_PROFILES_LEN = 40


@router.get("", status_code=status.HTTP_200_OK)
@limiter.limit(threats_rate_limit)
def read_threats(
    request: Request,
    profiles: str = Query(
        default="",
        max_length=_MAX_PROFILES_LEN,
        description="Profils tenus, séparés par des virgules, ex. `P1a,P3`.",
    ),
) -> dict[str, Any]:
    """Rendre les 16 lignes, positionnées pour les profils demandés."""
    try:
        held = triage.parse_profiles(profiles) if profiles.strip() else frozenset()
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from None

    try:
        report = triage.diagnose(_MAP, held)
    except triage.MapUnavailable:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Relevé momentanément indisponible",
        ) from None

    return {
        "profiles": sorted(p.value for p in held),
        "cap": report.cap,
        # Les comptes viennent du moteur, pas d'un `filter` côté page. « Sur notre
        # terrain » en particulier : `Diagnostic.ours` est défini une fois pour que
        # deux lecteurs n'en donnent pas deux chiffres.
        "counts": {
            "lines": len(report.lines),
            "applicable": len(report.applicable),
            "ours": len(report.ours),
            "blocked": len(report.blocked),
        },
        # Vide tant qu'aucun profil n'est coché : l'énoncé de `THREAT-COVERAGE` §2.5
        # parle d'« un client », et sans profil il n'y a pas de client à qui parler.
        "statement": report.statement() if held else None,
        "rows": [
            {
                "id": line.id,
                "titre": line.titre,
                "applicable": line.applicable,
                "blocked": line.blocked,
                "owner": line.owner,
                "facets": [{"libelle": libelle, "mode": mode} for libelle, mode in line.facets],
                "activates_at": [p.value for p in line.activates_at],
            }
            for line in report.lines
        ],
    }

File: frontend/lib/threats.ts
/**
 * Le relevé des menaces, côté front (`L3`).
 *
 * Deux moitiés, et la frontière entre elles est la règle du chantier.
 *
 * **Ce qui ne dépend d'aucun visiteur** — les 16 lignes, leurs facettes, leur mode
 * publié, les profils qu'elles concernent — vient de `generated/threat-rows.json`,
 * produit par `scripts/gen_threat_rows.py` depuis `coverage/map.json` et gaté en CI.
 * Rien n'est écrit à la main ici, et rien ne peut l'être : le fichier est régénéré.
 *
 * **Ce qui dépend du visiteur** — ce qui le concerne, ce que son plafond de profil
 * rabat, à qui revient le reste, et surtout les *comptes* — se demande à
 * `/api/threats`. Le recalculer ici donnerait un second moteur, et le premier
 * désaccord entre les deux se lirait sur une page commerciale. L'écart n'est pas
 * théorique : sur le profil `P1a`, la carte compte 3 facettes bloquées et 2 lignes
 * bloquées. Le nombre à écrire à côté du mot « ligne » est 2.
 */

import donnees from "@/lib/generated/threat-rows.json";

/** Les modes de couverture publiables. `NA` dit qu'aucune revendication n'est faite. */
export type Mode = "B" | "D" | "O" | "A" | "X" | "NA";

/** À qui le problème appartient. */
export type Famille = "fournisseur" | "cyber_classique" | "usage_ia";

export type ProfilId = "P1a" | "P1b" | "P2" | "P3" | "P4" | "P5";

export interface Facette {
  cle: string;
  libelle: string;
  mode: string;
  famille: string;
  profils: string[];
  ingress_prouve: string[];
  ingress_non_asserte: string[];
  scenarios: string[];
  ecarts: string[];
  raison: string | null;
}

export interface LigneMenace {
  id: string;
  titre: string;
  /** La ligne porte-t-elle de quoi être **montrée** plutôt qu'affirmée. */
  rejouable: boolean;
  profils: string[];
  facettes: Facette[];
}

export interface Releve {
  /** Le commit de la carte dont ce relevé est tiré. */
  commit: string;
  genere_le: string;
  lignes: LigneMenace[];
}

// Assigné **sans cast**, et c'est délibéré : `as Releve` n'aurait rien vérifié.
// Ainsi, `tsc` confronte structurellement le JSON généré aux types ci-dessus, et la
// CI rougit si `scripts/gen_threat_rows.py` change de forme sans que ce fichier suive.
export const RELEVE: Releve = donnees;

/** Ce que renvoie `/api/threats` : les chiffres, calculés par le moteur. */
export interface ReleveProfil {
  profiles: string[];
  cap: string | null;
  counts: { lines: number; applicable: number; ours: number; blocked: number };
  /** L'énoncé de `THREAT-COVERAGE` §2.5, ou `null` tant qu'aucun profil n'est coché. */
  statement: string | null;
  rows: {
    id: string;
    titre: string;
    applicable: boolean;
    blocked: boolean;
    owner: string;
    facets: { libelle: string; mode: string }[];
    activates_at: string[];
  }[];
}

/**
 * Demander le relevé positionné.
 *
 * Ne rattrape pas l'erreur : l'appelant décide quoi montrer, et une valeur de repli
 * inventée ici deviendrait une couverture affirmée sans preuve.
 */
export async function fetchReleveProfil(profils: readonly string[]): Promise<ReleveProfil> {
  const res = await fetch(`/api/threats?profiles=${encodeURIComponent(profils.join(","))}`);
  if (!res.ok) {
    const corps = (await res.json().catch(() => ({}))) as { detail?: string };
    throw new Error(corps.detail ?? "Relevé indisponible");
  }
  return (await res.json()) as ReleveProfil;
}
